delivery_carrier_label_gls/models/stock_picking.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 ACSONE SA/NV
+# © 2015 David BEAL @ Akretion
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import logging
+
+from odoo import _, fields, models
+from odoo.exceptions import ValidationError
+
+_logger = logging.getLogger(__name__)
+
+
+def joincat(s1, s2):
+    """Simply append a ref to a string, or get the string itself if s1 is falsy."""
+    return s2 if not s1 else ",".join((s1, s2))
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    gls_parcel_shop = fields.Char(
+        "GLS Parcel Shop Identifier",
+        help="Fill this for a delivery to a ParcelShop.",
+    )
+    gls_package_ref = fields.Char("GLS Package Identifiers", readonly=True, copy=False)
+
+    def _check_is_everything_packaged(self):
+        for picking in self:
+            if not all(
+                o.result_package_id for o in picking.move_ids.mapped("move_line_ids")
+            ):
+                msg = _("For GLS every operation should be put in a pack.")
+                raise ValidationError(msg)
+
+    def button_validate(self):
+        """Check that each GLS picking has been completely sent."""
+        gls_pickings = self.filtered(lambda p: p.delivery_type == "gls")
+        gls_pickings._check_is_everything_packaged()
+        return super().button_validate()
+
+    def gls_send_shipping(self, delivery_carrier=False):
+        for package in self.move_ids.mapped("move_line_ids.result_package_id"):
```

Comment:
mapped can be dropped
```suggestion
        for package in self.move_ids.move_line_ids.result_package_id:
```